### ISSUE TITLE
Start storybook with limited (filtered) stories

### ISSUE THREAD

**Issue by phaistonian** [storybookjs/storybook#5065]
**Is your feature request related to a problem? Please describe.**
Yes.

Having a lot of (complex) stories can make recurring updates painful since they can take a long, long time.

**Describe the solution you'd like**
Most of the times you only care for a subset of stories  —  if not just one.

How about passing an expression when starting storybook that will be used to filter out what stories load and work with?

For example:

- Before: `yarn storybook`
- After: `yarn storybook myapp`

As I noticed this, this would bring huge gains in some cases in terms of DX and sanity.

**Describe alternatives you've considered**
None

**Are you able to assist bring the feature to reality?**
Well, I guess it's not hard.

You need to pass vars from cli to webpack and then from there to `config` file (`DefinePlugin`?).

Then, use the argument passed to either narrow down context or filter out the matched files.

**Additional context**
Check out attachment:

![image](https://user-images.githubusercontent.com/796530/50266157-41780400-042b-11e9-96cf-60d70ca48397.png)

**Comment by backbone87** [comment]
if loading time is a no issue, this could be done internally after reading all stories, the storystore could just filter over all stories (by kind and story)

**Comment by phaistonian** [comment]
Yes, that's not an issue.

What is, however, is recurring updates.

**Comment by gearsandcode** [comment]
@phaistonian - did you ever get this sorted out?

**Comment by shilman** [comment]
Missed this the first time. Great idea if somebody wants to pick it up!

**Comment by kohakukun** [comment]
Would it be sufficient to pass all arguments from the cli to the story loader?
For example
`start-storybook -p 9010 -s public --filter someExpression`

....

```
import { configure } from '@storybook/react';

configure((command)=>{
  console.log(command.p) // 9010
  console.log(command.filter) // someExpression
 ...
}, module)
```

**Comment by joshuatvernon** [comment]
I got it to work like this . . .

### config.ts
```
const req = require.context('../src', true, /.stories.tsx$/);

const loadStories = () => {
  let keys = req.keys();
  if (process.env.STORYBOOK_FILTER) {
    keys = keys.filter(key => key.includes(process.env.STORYBOOK_FILTER));
  }
  keys.forEach(req);
};

configure(loadStories, module);
```

### package.json
```
...
"scripts": {
    ...
    "storybook": "startStorybook() { STORYBOOK_FILTER=\"$1\" start-storybook -p 4002 -c .storybook; }; startStorybook",
    ...
  },
...
```

**Comment by phaistonian** [comment]
@joshuatvernon Thing is, this way the `require.context` has already done its thing, plus this is working - in terms of filtering - only for the non-CSF format.

I managed to do it in the webpack level - so it's irrelevant with the story format - in a rather ugly way that involves using `ContextReplacementPlugin`.

Here's the code block:

```js
  // If we have passe an extra argument to `yarn storybook`, we shall take it and use to filter context
  if (process.env.npm_config_argv) {
    const argv = JSON.parse(process.env.npm_config_argv);

    if (argv && argv.original) {
      const filters = argv.original.slice(1);

      if (!filters.length) {
        console.log(
          '\n\nTip: You can try %s to filter for stories based on this filter\n\n',
          chalk.cyan.underline('yarn storybook input input2')
        );
      }

      const filter = filters[0];

      const isLimitingToPath = filter && filter[0] === '/';
      const shouldStrict = filter && filter[filter.length - 1] === '.';

      if (shouldStrict) {
        filter = filter.slice(0, -1);
      }

      if (filter && !isLimitingToPath) {
        const regString =
          filters.length === 1 ? filter : `(${filters.join('|')})`;

        console.log('\n🔍 Filtering stories by:', chalk.cyan(regString));

        // stories only
        const reg = shouldStrict
          ? new RegExp(`${regString}.stories.js$`, 'gi')
          : new RegExp(`${regString}.+?stories.js$`, 'gi');

        newConfig.plugins.push(
          new webpack.ContextReplacementPlugin(
            /\/(js|page|another_path)/,
            reg
          )
        );
      }
    }
  }
```

This way I can do:

```
yarn storybook story1 story 2

yarn storybook story.*?blue
``` 

You need to make sure the regexp passed to the `ContextReplacementPlugin` matches the paths where your stories are.

ps: the code is ugly, but you get the idea :)

**Comment by joshuatvernon** [comment]
Hmm I tried to get this to work but haven't been able to. I'm using Typescript so it may be a bit different.

This is what I've got so far...
```
const webpack = require("webpack");

const createStyledComponentsTransformer = require('typescript-plugin-styled-components').default;
const styledComponentsTransformer = createStyledComponentsTransformer();

module.exports = ({ config }) => {
  let pattern;
  const argv = JSON.parse(process.env.npm_config_argv);
  if (argv && argv.original) {
    const filters = argv.original.slice(2, argv.original.length);
    if (!filters.length) {
      console.log('\n\nTip: You can try npm storybook story1 story2 to filter stories\n\n');
      pattern = /\.(ts|tsx)$/;
    } else {
      // https://regex101.com/r/GfOQ6b/1
      pattern = new RegExp(`.*(?<=(${filters.join('|')}).*stories.*|(?<!stories.*))(\.(ts|tsx)$)`);
      console.log(pattern);
    }
  } else {
    pattern = /\.(ts|tsx)$/;
  }

  config.plugins.push(
    new webpack.ContextReplacementPlugin(
      /\.(ts|tsx)$/,
      pattern
    )
  );

  config.module.rules.push({
    test: /\.(ts|tsx)$/,
    use: [
      {
        loader: require.resolve('awesome-typescript-loader'),
        options: {
          presets: [['react-app', { flow: false, typescript: true }]],
          configFileName: './.storybook/tsconfig.json',
          getCustomTransformers: () => ({ before: [styledComponentsTransformer] })
        }
      },
      {
        loader: require.resolve('react-docgen-typescript-loader')
      }
    ]
  });

  config.resolve.extensions.push('.ts', '.tsx');

  return config;
};
```

**Comment by YozhEzhi** [comment]
**For monorepo with packages only.**

You could run all stories in package but not in all repo.

```
// package.json

{
    "name": "my-repo",
    "scripts": {
        "start": "start-storybook -p 9001 -c .storybook",
        "package:start": "STORYBOOK_STORYPATH=\"$INIT_CWD/**/*.stories.@(mdx|ts|tsx)\" && yarn start",
    }
}
```

```
// packages/my-package/package.json

{
    "name": "my-package",
    "scripts": {
        "start": "yarn package:start"
    }
}
```

```
// .storybook/main.js

module.exports = {
    stories: process.env.STORYBOOK_STORYPATH
        ? [process.env.STORYBOOK_STORYPATH] // <-- runs only package stories
        : ['../packages/**/*.stories.@(js|jsx|mdx|ts|tsx)'], // <-- runs all stories in repo
};
```

**Comment by ndelangen** [comment]
The modern way of configuring storybook uses a lost of globs.

You can respond to environment variables when setting the globs too. Therefore this issue is resolved.
